src/libraries/System.Private.CoreLib/src/System/Number.NumberToFloatingPointBits.cs

```diff
@@ -764,5 +1534,239 @@ private static bool ShouldRoundUp(bool lsbBit, bool roundBit, bool hasTailBits)
+                    answer.mantissa = 0;
+                    // result should be zero
+                    return answer;
+                }
+                // next line is safe because -answer.power2 + 1 < 64
+                answer.mantissa >>= -answer.power2 + 1;
+                // Thankfully, we can't have both "round-to-even" and subnormals because
+                // "round-to-even" only occurs for powers close to 0.
+                answer.mantissa += (answer.mantissa & 1); // round up
+                answer.mantissa >>= 1;
+                // There is a weird scenario where we don't have a subnormal but just.
+                // Suppose we start with 2.2250738585072013e-308, we end up
+                // with 0x3fffffffffffff x 2^-1023-53 which is technically subnormal
+                // whereas 0x40000000000000 x 2^-1023-53  is normal. Now, we need to round
+                // up 0x3fffffffffffff x 2^-1023-53  and once we do, we are no longer
+                // subnormal, but we can only know this after rounding.
+                // So we only declare a subnormal if we are smaller than the threshold.
+                answer.power2 = (answer.mantissa < ((ulong)(1) << info.DenormalMantissaBits)) ? 0 : 1;
+                return answer;
+            }
+
+            // usually, we round *up*, but if we fall right in between and and we have an
+            // even basis, we need to round down
+            // We are only concerned with the cases where 5**q fits in single 64-bit word.
+            if ((product.low <= 1) && (q >= info.MinExponentRoundToEven) && (q <= info.MaxExponentRoundToEven) &&
+                ((answer.mantissa & 3) == 1))
+            { // we may fall between two floats!
+              // To be in-between two floats we need that in doing
+              //   answer.mantissa = product.high >> (upperbit + 64 - mantissa_explicit_bits() - 3);
+              // ... we dropped out only zeroes. But if this happened, then we can go back!!!
+                if ((answer.mantissa << (upperbit + 64 - info.DenormalMantissaBits - 3)) == product.high)
+                {
+                    answer.mantissa &= ~(ulong)(1);          // flip it so that we do not round up
+                }
+            }
+
+            answer.mantissa += (answer.mantissa & 1); // round up
+            answer.mantissa >>= 1;
+            if (answer.mantissa >= ((ulong)(2) << info.DenormalMantissaBits))
+            {
+                answer.mantissa = ((ulong)(1) << info.DenormalMantissaBits);
+                answer.power2++; // undo previous addition
+            }
+
+            answer.mantissa &= ~((ulong)(1) << info.DenormalMantissaBits);
+            if (answer.power2 >= info.InfinitePower)
+            { // infinity
+                answer.power2 = info.InfinitePower;
+                answer.mantissa = 0;
+            }
+            return answer;
+        }
+
+
+        private static int LeadingZeroCount(ulong value)
```

Comment:
There's already https://github.com/dotnet/runtime/blob/12a8819eee9865eb38bca6c05fdece1053102854/src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs#L211 inbox with softwarefallback if intrinsics aren't supported, so just use `BitOperations.LeadingZeroCount`.

Reply:
Also, I guess `HAS_BITOPERATIONS`, `NET5_0_OR_GREATER` and `HAS_INTRINSICS` are not needed anymore